### PR TITLE
add metric for grpc 

### PR DESCRIPTION
**Pull request by wenxuwan**
### Issues associated with this PR

add metric for grpc request

### Solutions

add stream filter for grpc metric

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by wenxuwan** [comment]
@nejisama @zhenjunMa  please help review it.

File: pkg/config/v2/filter.go [hunk 1]
```diff
@@ -101,6 +101,7 @@ const (
 	DubboStream                = "dubbo_stream"
 	JwtAuthn                   = "jwt_authn"
 	GoPluginStreamFilterSuffix = "so_plugin"
+	GRPCStreamFilter           = "grpc"
```
Comment (nejisama) [inline pkg/config/v2/filter.go:104]: 这个改个名字，别叫grpc 的stream filter了，应该是grpc的metrics，grpc这个名字太宽泛了。stream filter的目录也改一下名字

File: pkg/filter/network/grpc/var.go [hunk 2]
```diff
@@ -23,13 +23,19 @@ import (
...
+	RequestResult string = "requestResult"
+)
+
+const (
+	SUCCESS string = "success"
+	FAIL    string = "fail"
 )
 
 var (
 	builtinVariables = []variable.Variable{
```
Comment (nejisama) [inline pkg/filter/network/grpc/var.go:37]: 这些变量应该定义在streamfilter里吧？

File: pkg/filter/network/grpc/var.go [hunk 3]
```diff
@@ -23,13 +23,19 @@ import (
...
 
 const (
-	grpcName    string = "gRPC"
-	serviceName string = "serviceName"
+	grpcName      string = "gRPC"
+	ServiceName   string = "serviceName"
+	RequestResult string = "requestResult"
+)
+
+const (
```
Comment (nejisama) [inline pkg/filter/network/grpc/var.go:31]: 这些应该定义在stream filter里吧？是metrics的一部分

File: pkg/filter/network/grpc/var.go [hunk 4]
```diff
@@ -23,13 +23,19 @@ import (
 )
 
 const (
-	grpcName    string = "gRPC"
-	serviceName string = "serviceName"
+	grpcName      string = "gRPC"
+	ServiceName   string = "serviceName"
+	RequestResult string = "requestResult"
```
Comment (nejisama) [inline pkg/filter/network/grpc/var.go:28]: 这个也应该是在stream filter里的，逻辑尽量不要到处耦合

File: pkg/filter/stream/grpc/stats.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	ServiceReqNum = "request_total"
+	ResponseSucc  = "response_succ_total"
+	ResponseFail  = "response_fail_total"
+	serviceKey    = "service"
+	metricPre     = "grpc"
+)
+
+var (
+	mux          sync.RWMutex
+	statsFactory = make(map[string]*Stats)
```
Comment (nejisama) [inline pkg/filter/stream/grpc/stats.go:21]: 这里就不要用全局变量了吧，变成stream filter factory的一个成员？

**Review by nejisama — COMMENTED** [review]
补充一下单元测试，代码里一点单元测试都没有

File: pkg/config/v2/filter.go (hunk 1 above)
Comment (wenxuwan) [reply pkg/config/v2/filter.go]: done

File: pkg/filter/network/grpc/var.go (hunk 3 above)
Comment (wenxuwan) [reply pkg/filter/network/grpc/var.go]: 这些定义主要是grpc的拦截器也需要用到，我放到types里面吧，这样net filter和stream filter都依赖types

File: pkg/filter/network/grpc/var.go (hunk 4 above)
Comment (wenxuwan) [reply pkg/filter/network/grpc/var.go]: 和上面一样的做法

File: pkg/filter/stream/grpc/stats.go (hunk 5 above)
Comment (wenxuwan) [reply pkg/filter/stream/grpc/stats.go]: 这个地方包内的全局变量应该还好吧，如果做factory的成员变量，filter拿不到吧？

**Comment by wenxuwan** [comment]
> 补充一下单元测试，代码里一点单元测试都没有

已加

File: pkg/filter/stream/grpc/stats.go (hunk 5 above)
Comment (nejisama) [reply pkg/filter/stream/grpc/stats.go]: filter是你的 factory创建的，为什么会拿不到呢，要减少全局变量的使用

File: pkg/filter/network/grpc/var.go (hunk 3 above)
Comment (nejisama) [reply pkg/filter/network/grpc/var.go]: 没理解这里为什么存在耦合关系，放到types里也是不合适的 应该要看下是不是设计有问题了。 相关的信息应该通过变量（ctx）传递 而不是去Header里进行修改

File: pkg/filter/network/grpc/factory.go [hunk 6]
```diff
@@ -167,12 +167,15 @@ func (f *grpcServerFilterFactory) UnaryInterceptorFilter(ctx context.Context, re
 
 	//do biz logic
 	resp, err = handler(newCtx, req)
-
 	responseHeader := header.CommonHeader{}
 	for k, v := range wrapper.header {
 		responseHeader.Set(k, v[0])
 	}
-
+	responseHeader.Set(types.GrpcServiceName, info.FullMethod)
```
Comment (nejisama) [inline pkg/filter/network/grpc/factory.go:174]: 这种修改Header来传递信息的行为不应该再存在了，应该使用变量进行传递

File: pkg/types/grpc.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package types
+
+const (
```
Comment (nejisama) [inline pkg/types/grpc.go:3]: 这些可以作为grpc filter的变量，存在grpc 的变量里

File: pkg/types/grpc.go [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+	GrpcName          string = "gRPC"
+	GrpcServiceName   string = "serviceName"
+	GrpcRequestResult string = "requestResult"
+)
+
+const (
+	SUCCESS string = "success"
```
Comment (nejisama) [inline pkg/types/grpc.go:10]: 使用变量的场景，这些可以直接转换成bool 类型，没有必要再定义这些东西了，这样metrics的代码和grpc的代码就没有那么多耦合了

Comment (wenxuwan) [reply pkg/types/grpc.go]: done

File: pkg/types/grpc.go (hunk 7 above)
Comment (wenxuwan) [reply pkg/types/grpc.go]: done

File: pkg/filter/network/grpc/factory.go (hunk 6 above)
Comment (wenxuwan) [reply pkg/filter/network/grpc/factory.go]: done

File: pkg/filter/stream/grpc/stats.go (hunk 5 above)
Comment (wenxuwan) [reply pkg/filter/stream/grpc/stats.go]: done

File: pkg/filter/network/grpc/var.go (hunk 3 above)
Comment (wenxuwan) [reply pkg/filter/network/grpc/var.go]: done

File: pkg/filter/network/grpc/factory.go [hunk 9]
```diff
@@ -150,11 +151,11 @@ func (f *grpcServerFilterFactory) UnaryInterceptorFilter(ctx context.Context, re
 		}
 	}
 
-	ctx = mosnctx.WithValue(ctx, types.ContextKeyDownStreamProtocol, api.ProtocolName(grpcName))
+	ctx = mosnctx.WithValue(ctx, types.ContextKeyDownStreamProtocol, api.ProtocolName(GrpcName))
 	ctx = mosnctx.WithValue(ctx, types.ContextKeyDownStreamHeaders, requestHeader)
 	ctx = variable.NewVariableContext(ctx)
 
-	variable.SetString(ctx, grpcName+"_"+serviceName, info.FullMethod)
+	variable.SetString(ctx, GrpcName+"_"+GrpcServiceName, info.FullMethod)
```
[the remark this answers is not in the document]
Comment (wenxuwan) [reply pkg/filter/network/grpc/factory.go:159]: done

**Review by taoyuanyuan — DISMISSED** [review]
lgtm

File: pkg/filter/network/grpc/factory.go [hunk 10]
```diff
@@ -150,50 +152,52 @@ func (f *grpcServerFilterFactory) UnaryInterceptorFilter(ctx context.Context, re
...
 
 	status := ss.RunReceiverFilter(ctx, api.AfterRoute, requestHeader, nil, nil, ss.receiverFilterStatusHandler)
 	if status == api.StreamFiltertermination || status == api.StreamFilterStop {
-		return nil, ss.err
+		err = ss.err
+	} else {
+		stream := grpc.ServerTransportStreamFromContext(ctx)
```
Comment (nejisama) [inline pkg/filter/network/grpc/factory.go:165]: 这里逻辑调整一下，尽量对标mosn的proxy，以前这里就有点问题。

Comment (wenxuwan) [reply pkg/filter/network/grpc/factory.go:170]: done，StreamFiltertermination 看了下很少会返回这个错误，但还是按照proxy逻辑修改代码

**Review by nejisama — APPROVED** [review]

**Review by taoyuanyuan — APPROVED** [review]
lgtm
